### DISCUSSION TITLE
Properties window layout

### DISCUSSION THREAD

**Discussion by feetstv** [obsproject/obs-studio#6726] [General]
Hello, OBS friends.

I would like to propose a change to the venerable Properties window for sources. It is, and always has been, my belief that the Properties window does not make efficient use of screen real estate, particularly on smaller displays. To my mind, this comes down to the vertical layout of the window. The issues are:

- to display the source preview at a reasonable size, the window must be fairly large **or** the splitter must be dragged down, thus showing few property widgets. This leaves laptop users with two choices:
  - small preview with a lot of property widgets
  - reasonable preview with few property widgets
- to display the source preview at a reasonable *width*, the window must be fairly wide. This elongates the property widgets — in most cases, they don't need to be that long

# Proposal

This proposal comes in two parts:

1. Make the Properties window layout horizontal and fix the Properties view width to 300px
2. Update the Properties view layout so that widgets are smaller and laid out more space-efficiently

## Make the Properties window layout horizontal

The motivation here is to make better use of the window's real estate.

At a given sensible, preview at the left and Properties view at the right, there *should* be more property widgets be visible than in the current layout.

Also, the Properties view should be fixed in width to around 300px (subject to testing). This means:

- the layout is easier to use than using a splitter whilst solving the issue that the splitter was intended to resolve
  - the size of the preview is contingent on the window size, not window size **and** splitter
- this design lends the Properties view to being repurposable as a dock in future

## Update the Properties view layout

To accommodate the Properties view being located on the side within a fixed width view, properties view widgets should be made smaller. The most straight-forward way to do this is to use a smaller font size.

Currently, the Properties view uses a QFormLayout to manage all of the widgets. This makes all labels for each widget the same width as the longest label. This is an incredibly wasteful use of horizontal screen real estate:

- if only one label is very long, *every* label becomes long, wasting horizontal space
- in the current vertical layout, some labels require the Properties window to be *very* wide in order to accommodate both the label and the widget
- in the horizontal layout, the Properties view would be so wide as to defeat the purpose of the change

Instead of using a QFormLayout, I propose a QVBoxLayout. This would change the layout from:

```
     Label: [     Widget     ]
Long label: [     Widget     ]
```
to
```
Label:
[     Widget     ]
Long label:
[     Widget     ]
```

Some widgets could further be adapted. For example:

- buttons with consistent text (Pick Color, Open File, etc.) could be reduced to a button with a glyph
  - these smaller buttons could be placed in-line with the label.
- superfluous or unnecessary labels could be reworded, restyled, or removed
- the Pick Color button in particular could be changed to simulate a color well (such as on web browsers with the ``<input type='color'>`` tag or macOS' NSColorWell control)

For example, the color picker could be changed so that:

1. the color code label is removed
2. the Pick Color button's text is removed
4. the Pick Color button's background color is changed to match the selected color

```
Color #1:
[     #ffffff     ] [Pick Color]
```
to the more space-saving
```
Color #1:    [   ]
```

# Example

The result of such changes are the following:

### Video capture source

![Video capture source](https://user-images.githubusercontent.com/88599935/178687514-7614e484-e3fe-4d64-83b7-f364c6f4055a.png)

### Image source

![Image source](https://user-images.githubusercontent.com/88599935/178688054-8d588a09-73cd-41a9-a3fe-83e55c347506.png)

### Media source

![Media source](https://user-images.githubusercontent.com/88599935/178687913-e695dd15-9937-4209-835b-f523decb5f7f.png)

### Display capture source

![Display capture source](https://user-images.githubusercontent.com/88599935/178687747-fcc1d537-4109-4fdd-ad7d-e2a353778853.png)

### Text source

Note the color well-like buttons.

![Text source](https://user-images.githubusercontent.com/88599935/178688143-ecca06aa-0bc7-405b-8d54-2dd8f856b2c8.png)

### Browser source

![Browser source](https://user-images.githubusercontent.com/88599935/178688208-bdeffefe-c48d-43df-b344-9a3fe0189d69.png)

### Browser source (small window)

![Small Browser source](https://user-images.githubusercontent.com/88599935/178688285-4df7cc1c-1df9-4bee-9917-699a42d0f946.png)

# Prior art

Narrow properties views, also called inspectors, are not uncommon. Apple (in its pro and productivity apps) and Adobe make good use of them, most software development IDEs have some kind of inspector layout.

![Layout demo 1](https://user-images.githubusercontent.com/88599935/178701511-f5093758-1d35-42e9-bc0a-b85eb17707b2.jpeg)

![Layout demo 2](https://user-images.githubusercontent.com/88599935/178701538-02cfcaaa-7a0d-4b28-8a65-ad6fbde6c71e.jpeg)

![Layout demo 3](https://user-images.githubusercontent.com/88599935/178701544-edddc496-59e2-4364-ac66-0edc1ff87eb5.jpeg)

# Conclusion

As you can see, I have been working on code to achieve this. I need to clean it up before I make a PR. But before I do, I wanted to ask for comments or suggestions on how this could be improved.

### Origin of the proposal

@gxalpha was working on a Properties Dock, I forked it and made similar changes to what I describe below. However, I realised that there would always be a mismatch between my version of the Properties Dock with its dock-optimised layout and the regular Properties window.

I believe that what I am proposing would benefit both the current workflow using a non-modal window *and* any future version of a Properties Dock which would benefit from being optimised for the narrow layout.

**Comment by WizardCM** [comment]
I haven't read your whole proposal yet, but I wanted to mention - Filters show the lists of filters on a sidebar on the left. Would it feel weird having that on the left and this on the right?

**Reply by feetstv** [reply]
This proposal is just for the Properties Window, not for Filters. I haven't quite figured that window out yet — perhaps filters list between the preview and the Properties view? Not sure.

I purposefully didn't mention it to keep the scope as focused as possible.

**Reply by tytan652** [reply]
Note that encoder in advanced mode also use the properties API (what is used to show those properties).

Edit: but we can just add a boolean to trigger the "newline" after the name of the option.

**Reply by WizardCM** [reply]
Sure, I mean more that if we were to make this change in a vacuum, users would be taught properties on the right, lists on the left. Is that a direction we want to persist?

Additionally, we would like to consolidate (at some stage) Properties and Filters into one window.

I do agree Filters is a tricky one due to the additional UI.

**Reply by feetstv** [reply]
> Note that encoder in advanced mode also use the properties API (what is used to show those properties).

Yes, and there's a couple of hiccups with that. I haven't shown the screenshots of what that looks like yet because it's early days from a technical standpoint. I really only made this for me alone; I got encouragement to put up the discussion from another contributor.

> I do agree Filters is a tricky one due to the additional UI.

That would be the ideal, possibly with tabs at the top to switch between Properties and Filters. Perhaps filter properties could be contained within some kind of accordion of collapsible sections, but I don't know what widget would be used for that in Qt.

Motion has the tab bars, Final Cut and Adobe's suite of apps have the collapsible sections.

**Reply by feetstv** [reply]
> Edit: but we can just add a boolean to trigger the "newline" after the name of the option

Yeah, I did something like that. The result is new style and old style co-existing peacefully.

<img width="1651" alt="QVBoxLayout and QFormLayout side-by-side" src="https://user-images.githubusercontent.com/747351/178853808-3e0e753e-2c0e-4f84-b043-cde993cbf6b2.png">

**Comment by feetstv** [comment]
I should point out, I'm aware of the massive spacing between the lines in the Properties view. Ironic of me to try to address screen real estate waste with … screen real estate waste! Obviously, that needs fixing, but you still get the general idea of the layout.

Ideally, the label for the widget would be closer.

**Comment by Fenrirthviti** [comment]
Is there a reason this was not submitted as an RFC? I understand that the RFCs have historically languished, but pushing further away from them is not something I would like to encourage at this time.

**Reply by feetstv** [reply]
I'm happy to submit as an RFC. Is there anything missing from this proposal that needs to be included in an RFC? If not, I can easily enough copy-paste the first post into an RFC.

**Reply by Fenrirthviti** [reply]
Nothing missing that I can see, I did just scan over it though. You'll want to use the RFC template, but I'm pretty sure the post covers all of those talking points already. Sorry for the hassle!

**Comment by feetstv** [comment]
Now posted as an [RFC](https://github.com/obsproject/rfcs/pull/50).

**Comment by Fenrirthviti** [comment]
Locking this as all further discussion should happen on the RFC linked above. Thanks for submitting!
